11783 failure with indexed relation

dbannik opened the pull request:
Fix issue https://github.com/doctrine/orm/issues/11783

bobvandevijver committed:
Fix Hydration when use ManyToMany[indexBy]

The bug related (#11694) and fixed mapping of sql column alias to field in entity (#11783) and
invalidate cache [cache/persisted/entity|cache/persisted/collection] when sql filter changes

greg0ire commented on src/Cache/CollectionCacheKey.php:
```diff
@@ -43,14 +43,14 @@ class CollectionCacheKey extends CacheKey
      * @param array<string, mixed> $ownerIdentifier The identifier of the owning entity.
      */
-    public function __construct($entityClass, $association, array $ownerIdentifier)
+    public function __construct($entityClass, $association, array $ownerIdentifier, string $filterHash)
```
Let's avoid this breaking change.

```suggestion
    public function __construct($entityClass, $association, array $ownerIdentifier, string $filterHash = '')
```

greg0ire commented on src/Cache/DefaultCache.php:
```diff
@@ -282,7 +282,7 @@ private function buildCollectionCacheKey(
         }
 
-        return new CollectionCacheKey($metadata->rootEntityName, $association, $ownerIdentifier);
+        return new CollectionCacheKey($metadata->rootEntityName, $association, $ownerIdentifier, '');
```
We're in `DefaultCache.php`, so let's use the default value.

```suggestion
        return new CollectionCacheKey($metadata->rootEntityName, $association, $ownerIdentifier);
```

greg0ire commented:
Also, please squash the commits together (since you wouldn't revert one without the others, right?), and  improve the remaining commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#crafting-meaningful-commit-messages). Specifically, explain both the bug and the fix.

dbannik commented:
@greg0ire  The problem with phpstan checking still exists, it should be fixed in a separate commit

greg0ire commented:
See #11793

dbannik commented:
> See #11793

Rebased PR

greg0ire commented on src/Cache/CollectionCacheKey.php:
```diff
@@ -43,14 +43,14 @@ class CollectionCacheKey extends CacheKey
         $this->association     = (string) $association;
 
-        parent::__construct(str_replace('\\', '.', strtolower($entityClass)) . '_' . implode(' ', $ownerIdentifier) . '__' . $association);
+        parent::__construct(str_replace('\\', '.', strtolower($entityClass)) . '_' . implode(' ', $ownerIdentifier) . '__' . $association . '_' . $filterHash);
```
I don't think there should be a trailing underscore if `$filterHash` is an empty string. It's going to have an impact on all existing entries for no good reason, and also, I think it will look ugly.

dbannik replied:
Resolve

dbannik commented:
> > The bug related (#11694) and fixed mapping of sql column alias to field in entity (#11783) and
> > invalidate cache [cache/persisted/entity|cache/persisted/collection] when sql filter changes
> 
> if you are doing 2 things in one commit, that's bad. You should be doing 2 commits with clear messages, so that if one of them is wrong, it can be understood and reverted while the other stays.

these 2 things can't be separated
since separately the tests will fail

SenseException commented on src/Cache/Persister/Entity/AbstractEntityPersister.php:
```diff
@@ -611,12 +616,13 @@ public function refresh(array $id, $entity, $lockMode = null)
      * @return CollectionCacheKey
      */
-    protected function buildCollectionCacheKey(array $association, $ownerId)
+    protected function buildCollectionCacheKey(array $association, $ownerId, string $filterHash)
```
@greg0ire Is introducing a new argument to a protected method okay in a version-branch that's already released? It may break projects that are implementing the Abstract class.

greg0ire replied:
I didn't spot that one. You're right, it's a breaking change for extending classes, that suddenly need to have that extra argument, and for callers, that need to provide it.

@dbannik please resort to `func_get_args()`. The new argument should be added in the next major release.

dbannik replied:
@greg0ire  instead of `func_get_args` can we use default value ?
```suggestion
    protected function buildCollectionCacheKey(array $association, $ownerId, string $filterHash = '')
````

greg0ire replied:
That would still be a breaking change for extending classes (they would suddenly have incompatible signatures).

dbannik replied:
resolved

greg0ire commented:
Thanks @dbannik !
